src/odemis/driver/actuator.py

```diff
@@ -2179,3 +2182,601 @@ def terminate(self):
+        A helper function to check if current position is in focus range
+        :return: True if position in focus range, None otherwise
+        """
+        pos = self.position.value['z'] if pos is None else pos
+        # Add 1% margin for hardware slight errors
+        margin = (self._range[1] - self._range[0]) * 0.01
+        if (self._range[0] - margin) <= pos <= (self._range[1] + margin):
+            return True
+
+    @isasync
+    def moveAbs(self, pos):
+        """
+        Move the focus to the defined position in m for the Z axis. This is an
+        asynchronous method.
+        pos dict(string-> float): name of the axis and new position in m
+        returns (Future): object to control the move request
+        """
+        if not pos:
+            return model.InstantaneousFuture()
+        # check pos value is in range (except when it's = deactive)
+        if not self._isInRange(pos['z']) and not util.almost_equal(pos['z'],
+                                                                   self._metadata[model.MD_FAV_POS_DEACTIVE]['z'],
+                                                                   rtol=1e-3):
+            raise ValueError(
+                "Position %s for axis z outside of range %f->%f" % (pos['z'], self._range[0], self._range[1]))
+        # Prevent movement when parent rx != 0
+        self._checkParentRxRotation()
+
+        f = self.parent._createFuture()
+        f = self._executor.submitf(f, self._doMoveAbs, f, pos['z'])
+        return f
+
+    @isasync
+    def moveRel(self, shift):
+        """
+        Move the focus by the defined values in m for the Z axis. This is an
+        asynchronous method.
+        shift dict(string-> float): name of the axis and shift in m
+        returns (Future): object to control the move request
+        """
+        if not shift:
+            return model.InstantaneousFuture()
+        # Prevent movement if focus is in deactive position
+        if util.almost_equal(self.position.value['z'], self._metadata[model.MD_FAV_POS_DEACTIVE]['z'], rtol=1e-3):
+            raise ValueError("Cannot move while focus is not active")
+
+        self._checkMoveRel(shift)
+        if not self._isInRange(self.position.value['z'] + shift['z']):
+            raise ValueError("Relative movement would go outside of range")
+        # Prevent movement when parent rx != 0
+        self._checkParentRxRotation()
+
+        f = self.parent._createFuture()
+        f = self._executor.submitf(f, self._doMoveRel, f, shift)
+        return f
+
+    def adjustFocus(self, future, vector_value, rel):
+        """
+        Acts as an interface for the parent stage to call when it moves, and focus needs to be adjusted accordingly
+        :param vector_value: name of the axis 'z' and new position value
+        :param rel: whether it's a relative movement or absolute
+        """
+        # Only move focus when it's in active range
+        if not self._isInRange():
+            return
```

Comment:
Added this check again to prevent movement when lens is parked, I believe it makes sense to drop the movement than throwing exceptions of either:
- Relative movement is out of range (deactive + value)
- Absolute second movement after parent stage rx changed. 

Both of these shouldn't throw exceptions, they just shouldn't happen.